crates/rpc/src/jsonrpc/websocket/logic.rs

```diff
@@ -0,0 +1,500 @@
+//! See [the parent module documentation](super)
+use std::collections::HashMap;
+use std::ops::ControlFlow;
+
+use crate::jsonrpc::request::RawParams;
+use crate::jsonrpc::{RequestId, RpcRequest};
+use axum::extract::ws::{Message, WebSocket};
+use axum::extract::{State, WebSocketUpgrade};
+use axum::response::IntoResponse;
+use futures::sink::Buffer;
+use futures::stream::{SplitSink, SplitStream};
+use futures::{SinkExt, StreamExt};
+use tokio::sync::mpsc;
+use tracing::error;
+
+use crate::jsonrpc::websocket::data::{Kind, ResponseEvent, SubscriptionId, SubscriptionItem};
+use crate::BlockHeader;
+
+const SUBSCRIBE_METHOD: &str = "pathfinder_subscribe";
+const UNSUBSCRIBE_METHOD: &str = "pathfinder_unsubscribe";
+const NEW_HEADS_TOPIC: &str = "newHeads";
+
+pub async fn websocket_handler(
+    ws: WebSocketUpgrade,
+    State(senders): State<WebsocketSenders>,
+) -> impl IntoResponse {
+    ws.on_upgrade(|socket| handle_socket(socket, senders))
+}
+
+async fn handle_socket(socket: WebSocket, senders: WebsocketSenders) {
+    let (ws_sender, ws_receiver) = socket.split();
+
+    let (msg_sender, msg_receiver) = mpsc::channel(100);
+
+    tokio::spawn(write(ws_sender, msg_receiver));
+    tokio::spawn(read(ws_receiver, msg_sender, senders));
+}
+
+async fn write(
+    sender: SplitSink<WebSocket, Message>,
+    mut msg_receiver: mpsc::Receiver<ResponseEvent>,
+) {
+    let mut sender = sender.buffer(100);
+    while let Some(response) = msg_receiver.recv().await {
+        if let ControlFlow::Break(()) = send_response(&mut sender, &response).await {
+            break;
+        }
+    }
+}
+
+async fn send_response(
+    sender: &mut Buffer<SplitSink<WebSocket, Message>, Message>,
+    response: &ResponseEvent,
+) -> ControlFlow<()> {
+    let message = match serde_json::to_string(&response) {
+        Ok(x) => x,
+        Err(e) => {
+            tracing::warn!(error=%e, kind=response.kind(), "Encoding websocket message failed");
+            return ControlFlow::Break(());
+        }
+    };
```

Comment:
A thought I had (and this is only applicable for general subscriptions like `newHeads`) - if we have many `newHeads` subscriptions, we technically only need to encode the header once and the result shared as `Arc<String>` to all the other senders. Does make things a lot more complex though I think..

Futher thought: is it not wasteful to encode here, and not together with the actual message output? Since we first allocate a string here, and then copy it again into the final message's json?

Reply:
I've found some partial success here by serializing the header to a `serde_json::Value` upon submitting it to the broadcast channel.
[This comparison](https://github.com/eqlabs/pathfinder/compare/bb0f631fbea7ae7bcfe735d2e4f784852f1ab836..a1cd784312e894946fbb821b32072acb9067c5a2) shows you my idea.
It won't cut all serialization costs as there's still the `Value`->bytes step that probably involves a fair amount of allocations, but `serde` makes it harder to get any further.

I'll close this thread but feel free to reopen if needed.